pep-0683.rst

```diff
@@ -0,0 +1,335 @@
+of the refcount will still match ``_Py_IMMORTAL_REFCNT`` so we can
+still identify such objects as immortal.
+
+No user-facing behavior changes, with the following exceptions:
+
+* code that inspects the refcount (e.g. ``sys.getrefcount()``
+  or directly via ``ob_refcnt``) will see a really, really large
+  value
+* ``Py_SET_REFCNT()`` will be a no-op for immortal objects
+
+Neither should cause a problem.
+
+Alternate Python Implementations
+--------------------------------
+
+This proposal is CPython-specific.
+
+Security Implications
+---------------------
+
+This feature has no known impact on security.
+
+Maintainability
+---------------
+
+This is not a complex feature so it should not cause much mental
+overhead for maintainers.  The basic implementation doesn't touch
+much code so it should have much impact on maintainability.  There
+may be some extra complexity due to performance penalty mitigation.
+However, that should be limited to where we immortalize all
+objects post-init and that code will be in one place.
+
+Non-Obvious Consequences
+------------------------
+
+* immortal containers effectively immortalize each contained item
+* the same is true for objects held internally by other objects
+  (e.g. ``PyTypeObject.tp_subclasses``)
+* an immortal object's type is effectively immortal
+* though extremely unlikely (and technically hard), any object could
+  be incref'ed enough to reach ``_Py_IMMORTAL_REFCNT`` and then
+  be treated as immortal
+
+
+Specification
+=============
+
+The approach involves these fundamental changes:
+
+* add ``_Py_IMMORTAL_REFCNT`` (the magic value) to the internal C-API
+* update ``Py_INCREF()`` and ``Py_DECREF()`` to no-op for objects with
+  the magic refcount (or its most significant bit)
+* do the same for any other API that modifies the refcount
+* ensure that all immortal objects are cleaned up during
+  runtime finalization
+
+Then setting any object's refcount to ``_Py_IMMORTAL_REFCNT``
+makes it immortal.
+
+To be clear, we will likely use the most-significant bit of
+``_Py_IMMORTAL_REFCNT`` to tell if an object is immortal, rather
+than comparing with ``_Py_IMMORTAL_REFCNT`` directly.
+
+(There are other minor, internal changes which are not described here.)
+
+This is not meant to be a public feature but rather an internal one.
+So the proposal does *not* including adding any new public C-API,
+nor any Python API.  However, this does not prevent us from
+adding (publicly accessible) private API to do things
+like immortalize an object or tell if one
+is immortal.
+
+Affected API
+------------
+
+API that will now ignore immortal objects:
+
+* (public) ``Py_INCREF()``
+* (public) ``Py_DECREF()``
+* (public) ``Py_SET_REFCNT()``
+* (private) ``_Py_NewReference()``
+
+API that exposes refcounts (unchanged but may now return large values):
+
+* (public) ``Py_REFCNT()``
+* (public) ``sys.getrefcount()``
+* (public) ``sys.gettotalrefcount()``
```

Comment:
This won't change! Or rather, it will change but not in the expected way.

This is something that I really looked into to make sure that the refleak tooling wouldn't break from this change. `gettotalrefcount ` relies on `_Py_RefTotal` which is updated independently from the object's refcount field. Thus, an object's immortality will not make the value of `gettotalrefcount` be something really large.

Currently, I've placed the `_Py_RefTotal` update after the immortalization check. Thus, this value now reflects all the increfs/decrefs that did take place. If we move the check to before the immortalization check, it will now reflect all the attempts to incref and decrefs that took place (but they could have been a no-op). I chose the former since it's a bit unintuitive to have a returned value that's larger than the actual effective operations. This is enough to make the refcount leak tooling to work as is. However, if we choose the latter, it will work even in cases where we can upgrade a once mortal object to be immortal (which is beyond of the scope of any of the current changes in the PR). In this case, `_Py_RefTotal` will always be updated regardless of the the immortalization status of the instance, but the returned value might make less sense as a whole.

The final consideration is that we need to make sure that `_PySet_Dummy` is not set to be immortal, otherwise, it will affect the implementation of `_Py_GetRefTotal`. Perhaps we should just remove the use of `_PySet_Dummy` and fix the gdb plugin in another way?

Reply:
I've mentioned that `sys.gettotalrefcount()` is unaffected but didn't include any of your explanation.  If you think any of it would be helpful to PEP readers then we can add it in a follow-up.